File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        int cnt=0;
        vector<vector<int>>vis(n,vector<int>(m,0));
        int dr[4]={-1,0,1,0};
        int dc[4]={0,1,0,-1};
        queue<pair<int,pair<int,int>>>q;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(!vis[i][j]){
                if(grid[i][j]=='1'){
                    cnt++;
                    vis[i][j]=1;
                    q.push({i,{j,0}});
                    while(!q.empty()){
                        int r=q.front().first;
                        int c=q.front().second.first;
                        int cnt=q.front().second.second;
                        q.pop();
                        // if(grid[n-1][m-1]) return cnt;
                        for(int i=0;i<4;i++){
                            int row=r+dr[i];
                            int col=c+dc[i];
                            if(row>=0 && row<n && col>=0 && col<m
                            && !vis[row][col] && grid[row][col]=='1'){
                                vis[row][col]=1;
                                q.push({row,{col,cnt+1}});
                            }
                        }
                    }
                }
            }
        }
    }
    return cnt;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
from collections import deque
class Solution:
    def numIslands(self, nums: List[List[str]]) -> int:
        n=len(nums)
        m=len(nums[0])
        vis=[[0]*m for _ in range(n)]
        q=deque()
        cnt=0
        dr=[1,0,-1,0]
        dc=[0,1,0,-1]
        for i in range(len(nums)):
            for j in range(len(nums[0])):
                if vis[i][j]==0 and nums[i][j]=='1':
                    vis[i][j]=1
                    cnt+=1
                    q.append((i,j))
                    while q:
                        r,c=q.popleft()
                        for k in range(4):
                            row=dr[k]+r
                            col=dc[k]+c
                            if row>=0 and row<n and col>=0 and col<m and vis[row][col]==0 and nums[row][col]=='1':
                                vis[row][col]=1
                                q.append((row,col))
        return cnt
